### ISSUE TITLE
NoSuchElementException in IR

### ISSUE THREAD

**Issue by InsanusMokrassar** [Kotlin/kotlinx.serialization#1940]
Currently in one of my [projects](https://git.inmo.dev/PostsSystem/core) I am getting `NoSuchElementException` error in compile time. Unfortunatelly, I am even do not know exactly which code is lead to the error.

 - Kotlin version:1.6.21
 - Library version: 1.3.3
 - Kotlin platforms: Multiplatform (compile error in JS IR)
 - Gradle version: 7.4.2 (wrapper)

**Comment by InsanusMokrassar** [comment]
Error stacktrace:

```bash
> Task :postssystem.features.auth.client:compileKotlinJs FAILED
e: java.util.NoSuchElementException: Collection contains no element matching the predicate.
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.createLazyProperty(GeneratorHelpers.kt:1419)
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableCompanionIrGenerator.createLazyProperty(SerializableCompanionIrGenerator.kt:35)
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableCompanionIrGenerator.generateLazySerializerGetter(SerializableCompanionIrGenerator.kt:114)
        at org.jetbrains.kotlinx.serialization.compiler.backend.common.SerializableCompanionCodegen.generate(SerializableCompanionCodegen.kt:37)
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableCompanionIrGenerator$Companion.generate(SerializableCompanionIrGenerator.kt:50)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializerClassLowering.lower(SerializationLoweringExtension.kt:60)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:36)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitClass(IrElementVisitorVoid.kt:44)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlin.ir.declarations.IrClass.accept(IrClass.kt:57)
        at org.jetbrains.kotlin.ir.declarations.IrClass.acceptChildren(IrClass.kt:62)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:287)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:37)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitClass(IrElementVisitorVoid.kt:44)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlin.ir.declarations.IrClass.accept(IrClass.kt:57)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.acceptChildren(IrFileImpl.kt:84)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:287)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitElement(SerializationLoweringExtension.kt:32)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitPackageFragment(IrElementVisitorVoid.kt:30)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitPackageFragment(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:37)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitFile(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:38)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitFile(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitFile(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.accept(IrFileImpl.kt:81)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:283)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt.runOnFileInOrder(SerializationLoweringExtension.kt:30)
        at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtension.generate(SerializationLoweringExtension.kt:77)
        at org.jetbrains.kotlin.ir.backend.js.KlibKt.generateModuleFragmentWithPlugins$lambda-23(klib.kt:590)
        at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment(Psi2IrTranslator.kt:99)
        at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment$default(Psi2IrTranslator.kt:75)
        at org.jetbrains.kotlin.ir.backend.js.KlibKt.generateModuleFragmentWithPlugins(klib.kt:595)
        at org.jetbrains.kotlin.ir.backend.js.KlibKt.generateIrForKlibSerialization(klib.kt:184)
        at org.jetbrains.kotlin.ir.backend.js.KlibKt.generateKLib(klib.kt:219)
        at org.jetbrains.kotlin.ir.backend.js.KlibKt.generateKLib$default(klib.kt:201)
        at org.jetbrains.kotlin.cli.js.K2JsIrCompiler.doExecute(K2JsIrCompiler.kt:269)
        at org.jetbrains.kotlin.cli.js.K2JSCompiler.doExecute(K2JSCompiler.java:178)
        at org.jetbrains.kotlin.cli.js.K2JSCompiler.doExecute(K2JSCompiler.java:71)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:91)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:43)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:93)
        at org.jetbrains.kotlin.incremental.IncrementalJsCompilerRunner.runCompiler(IncrementalJsCompilerRunner.kt:200)
        at org.jetbrains.kotlin.incremental.IncrementalJsCompilerRunner.runCompiler(IncrementalJsCompilerRunner.kt:77)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally(IncrementalCompilerRunner.kt:367)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally$default(IncrementalCompilerRunner.kt:309)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl$rebuild(IncrementalCompilerRunner.kt:115)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:167)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:77)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile$default(IncrementalCompilerRunner.kt:68)
        at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execJsIncrementalCompiler(CompileServiceImpl.kt:565)
        at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execJsIncrementalCompiler(CompileServiceImpl.kt:101)
        at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1807)
        at jdk.internal.reflect.GeneratedMethodAccessor104.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by altavir** [comment]
The erro comes from here: https://github.com/JetBrains/kotlin/blob/8de4eb798c67cc9231cb7425d90695800f40d2ab/plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/GeneratorHelpers.kt#L121-L124. But I don't know what those lines do.

**Comment by InsanusMokrassar** [comment]
[Here](https://git.inmo.dev/PostsSystem/core/src/commit/9216c013ece22dd0bfdfff3fa6f139663ef5122e) is the project at the state when it has compiler error.

[Here](https://git.inmo.dev/PostsSystem/core/src/commit/5bbe0b8a0e15f230caa2f8b7288d400626d4dc18) I have fixed the build via removing of `@Serializable` annotation for objects and replace sealed classes with sealed interfaces

**Comment by InsanusMokrassar** [comment]
[Here](https://git.inmo.dev/PostsSystem/core/src/commit/9216c013ece22dd0bfdfff3fa6f139663ef5122e) is the project at the state when it has compiler error.

[Here](https://git.inmo.dev/PostsSystem/core/src/commit/5bbe0b8a0e15f230caa2f8b7288d400626d4dc18) I have fixed the build via removing of `@Serializable` annotation for objects and replace sealed classes with sealed interfaces

**Comment by ankushg** [comment]
I'm also running into this error unexpectedly

The stacktrace/error message doesn't provide a whole lot to go off of, so I'm a bit stuck and unable to move forward 😔

**Comment by darussian** [comment]
Also Running into this problem on 1.7.0

**Comment by Reedyuk** [comment]
Also getting the same issue on my project. Is there anything else we need to provide to help?

**Comment by darussian** [comment]
@Reedyuk I was unable to find a solution other than what @InsanusMokrassar offered above, which is using sealed interfaces instead of classes.

**Comment by shanshin** [comment]
Hi, sorry for the delayed answer.

@InsanusMokrassar , could you rollback the changes to the state when the build fell with `NoSuchElementException` and insert such code into any file in the project where the error appears (eg `:postssystem.features.auth.client`)?

```
@Suppress("unused")
private val lazy = LazyThreadSafetyMode.PUBLICATION
```
*any use of the `LazyThreadSafetyMode.PUBLICATION` from the function body will also do, even in code that will never be executed.

**Comment by shanshin** [comment]
Fixed in `1.8.0`

**Comment by benkuly** [comment]
@shanshin How can I find the reason of the error to fix it until 1.8.0. I did a large refactoring, so there is not really a commit history to find the reason.

**Comment by InsanusMokrassar** [comment]
> Hi, sorry for the delayed answer.
> 
> @InsanusMokrassar , could you rollback the changes to the state when the build fell with `NoSuchElementException` and insert such code into any file in the project where the error appears (eg `:postssystem.features.auth.client`)?
> 
> ```
> @Suppress("unused")
> private val lazy = LazyThreadSafetyMode.PUBLICATION
> ```
> 
> *any use of the `LazyThreadSafetyMode.PUBLICATION` from the function body will also do, even in code that will never be executed.

Hello, I have tried this way (in the other project) and that fixed issue for me, thanks. Is there any chance to see this fix in `1.7.20`?

**Comment by sandwwraith** [comment]
Fixed in 1.8.0

**Comment by SettingDust** [comment]
> Hi, sorry for the delayed answer.
> 
> @InsanusMokrassar , could you rollback the changes to the state when the build fell with `NoSuchElementException` and insert such code into any file in the project where the error appears (eg `:postssystem.features.auth.client`)?
> 
> ```
> @Suppress("unused")
> private val lazy = LazyThreadSafetyMode.PUBLICATION
> ```
> 
> *any use of the `LazyThreadSafetyMode.PUBLICATION` from the function body will also do, even in code that will never be executed.

Can't fixed. I'm facing this issue when add `@Serializer` to my serializers for external polymorphic serialization.
